src/picongpu/include/plugins/particleDetector/ParticleDetectorFunctors.hpp

```diff
@@ -0,0 +1,208 @@
+/**
+ * Copyright 2016 Heiko Burau
+ *
+ * This file is part of PIConGPU.
+ *
+ * PIConGPU is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License as published by
+ * the Free Software Foundation, either version 3 of the License, or
+ * (at your option) any later version.
+ *
+ * PIConGPU is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with PIConGPU.
+ * If not, see <http://www.gnu.org/licenses/>.
+ */
+
+#pragma once
+
+#include "simulation_defines.hpp"
+#include "math/Vector.hpp"
+#include "algorithms/math.hpp"
+
+namespace picongpu
+{
+using namespace PMacc;
+
+template<typename CalorimeterCur>
+struct CalorimeterFunctor
+{
+    CalorimeterCur calorimeterCur;
+
+    const float_X maxTheta;
+    const float_X maxPhi;
+    const uint32_t numBinsTheta;
+    const uint32_t numBinsPhi;
+    const uint32_t numBinsEnergy;
+    const float_X minEnergy;
+    const float_X maxEnergy;
+    const bool logScale;
+
+    const float3_X detectorFrameVecX;
+    const float3_X detectorFrameVecY;
+    const float3_X detectorFrameVecZ;
+
+    CalorimeterFunctor(const float_X maxTheta,
+                     const float_X maxPhi,
+                     const uint32_t numBinsTheta,
+                     const uint32_t numBinsPhi,
+                     const uint32_t numBinsEnergy,
+                     const float_X minEnergy,
+                     const float_X maxEnergy,
+                     const bool logScale,
+                     const float3_X detectorFrameVecX,
+                     const float3_X detectorFrameVecY,
+                     const float3_X detectorFrameVecZ) :
+        calorimeterCur(NULL, PMacc::math::Size_t<DIM2>()),
+        maxTheta(maxTheta),
+        maxPhi(maxPhi),
+        numBinsTheta(numBinsTheta),
+        numBinsPhi(numBinsPhi),
+        numBinsEnergy(numBinsEnergy),
+        minEnergy(minEnergy),
+        maxEnergy(maxEnergy),
+        logScale(logScale),
+        detectorFrameVecX(detectorFrameVecX),
+        detectorFrameVecY(detectorFrameVecY),
+        detectorFrameVecZ(detectorFrameVecZ)
+    {}
+
+    HINLINE void setCalorimeterCursor(const CalorimeterCur& calorimeterCur)
+    {
+        this->calorimeterCur = calorimeterCur;
+    }
+
+    template<typename ParticlesFrame>
+    DINLINE void operator()(ParticlesFrame& particlesFrame, const uint32_t linearThreadIdx)
+    {
+        const float3_X mom = particlesFrame[linearThreadIdx][momentum_];
+        const float_X mom2 = math::dot(mom, mom);
+        float3_X dirVec = mom * math::rsqrt(mom2);
+
+        /* rotate dirVec into the detector frame */
```

Comment:
You are right, because your coordinate transformation will never consist of a translation. Could you still add `rotate/coordinate transformation` just to ease understanding.